kafka/pykafka/consumer.py

```diff
@@ -47,9 +199,63 @@ class OwnedPartition(object):
     Used to keep track of offsets and the internal message queue.
     """
 
-    def __init__(self, partition):
+    def __init__(self, partition, consumer):
         self.partition = partition
+        self.consumer = consumer
         self._messages = Queue()
+        self.last_offset_consumed = 0
+        self.next_offset = 0
+
+        if self.consumer._auto_commit_enable and self.consumer.consumer_group is not None:
+            self.last_offset_consumed = self._fetch_last_known_offset()
 
     def consume(self, timeout=None):
+        """Get a single message from this partition
+        """
+        if self._messages.empty():
+            self._fetch(timeout=timeout)
+
+        try:
+            message = self._messages.get_nowait()
+            self._queued_chunks -= 1
+            self.last_offset_consumed = message.offset
+            return message
+        except Empty:
+            return None
+
+    def _fetch_last_known_offset(self):
+        """Use the Offset Commit/Fetch API to find the last known offset for
+            this partition
+        """
         pass
+
+    def _fetch(self, timeout=None):
+        topic_name = self.partition.topic.name
+        success = False
+        while success is False:
+            try:
+                request = PartitionFetchRequest(
+                    self.partition.topic.name, self.partition.id, self.next_offset,
+                    self.consumer.fetch_message_max_bytes
+                )
+                response = self.partition.leader.fetch_messages(
+                    [request], timeout=timeout,
+                    min_bytes=self._fetch_min_bytes
+                )
+
+                messages = response.topics[topic_name].messages
+
+                if self._queued_chunks >= self._queued_max_message_chunks:
+                    return
```

Comment:
When changing this to messages queued: this doesn't have to be precise.  If you're supposed to stop at 100 queued messages, but you've downloaded 10 extra, there's no harm in adding them to the queue. The temporary bulk is preferable to redownloading those messages.